### PR TITLE
feat: multi-select messages in session chat

### LINKED ISSUE #978: Multi-select messages in session chat (WhatsApp-style)
## Summary

Add a WhatsApp/Telegram-style multi-select mode to the session event stream, with a hover action row on each event for quick single-message actions (copy, select). Users can select individual messages/events and bulk-copy their content as formatted text. This builds the selection infrastructure that the forwarding feature (#979) will extend.

## Parent

Part of #976 — Copy, multi-select, and forward messages in session chat

---

### Overview

The session chat UI currently has static copy buttons on messages and tool cards (added by #977), but no way to select or act on multiple messages. This ticket adds two layers: (1) a **hover action row** on each event for quick single-message actions (replacing the existing static copy button placement), and (2) a **multi-select mode** for bulk operations. The multi-select infrastructure lays the groundwork for the forward-to-session feature (#979).

### Functional Requirements

**Hover action row (single-message quick actions)**

- **FR-1**: When the user hovers over any event in EventStream, a small floating action row appears anchored to the event (top-right corner, similar to Slack/Discord message hover). On mobile, actions are accessible via long-press context menu instead.
- **FR-2**: The hover action row contains: **Copy** (copies that single event as markdown) and **Select** (enters multi-select mode with that event pre-selected). The row replaces the current static CopyButton placement on TextEvent and tool cards — individual copy buttons scattered across each event type are consolidated into this single, consistent hover affordance.
- **FR-3**: A right-click context menu on any event offers the same actions as the hover row (Copy, Select), providing a secondary entry point.
- **FR-4**: The hover row appears for all content-bearing event types: text/output, user_input, tool_use (settled), tool_result, error. It does NOT appear on status, signal, usage, or system events (these are metadata, not copyable content).

**Multi-select mode**

- **FR-5**: Clicking "Select" in the hover row (or right-click menu) activates selection mode and selects that event.
- **FR-6**: When selection mode is active, each content-bearing event shows a **checkbox** on the left side. The checkbox is a visual indicator; the entire event row is a click target for toggling selection.
- **FR-7**: **Shift+click** selects a range: all content-bearing events between the last-selected event and the shift-clicked event are selected (matching the file-explorer convention developers expect).
- **FR-8**: Selected events receive a subtle highlight background (e.g., primary color at 10% opacity).
- **FR-9**: Hover action rows are hidden while selection mode is active (the checkboxes replace them as the interaction surface).

**Floating action bar**

- **FR-10**: When selection mode is active, a floating action bar appears at the top of the EventStream. It contains: a selection count ("3 selected"), a **Select all / Deselect all** toggle, a **Copy** button, and a **Cancel** button.
- **FR-11**: The **Copy** button copies all selected events as formatted markdown (see FR-14) and shows a toast notification ("Copied N messages to clipboard"). Selection mode remains active after copy so the user can forward the same selection (#979).
- **FR-12**: The **Cancel** button (and the Escape key) exits selection mode and clears the selection.
- **FR-13**: The floating action bar is a placeholder for future actions. The Forward button will be added by #979.

**Copy format**

- **FR-14**: A formatEventsAsMarkdown(events: DisplayEvent[]) utility function produces well-formatted markdown from a list of events. The format is:

```
**User** (2:34 PM):
Can you fix the login bug?

**Assistant** (2:34 PM):
I will look into the authentication flow. Let me read the relevant files.

**Tool: Read** src/auth/login.ts
[file content]

**Shell** npm test
42 tests passed
```

- **FR-15**: The utility handles all content-bearing event types: user_input maps to "User", text/output maps to "Assistant", tool_result (with paired tool_use context) maps to "Tool: [name]" with args summary and result, error maps to "Error". Timestamps use the event timestamp field formatted as local time. Code blocks are preserved with fences. Tool cards are rendered as labeled blocks with their content.
- **FR-16**: When copying a single event via the hover row (FR-2), the same formatEventsAsMarkdown utility is used (with a single-element array), producing consistent output.

**Auto-scroll interaction**

- **FR-17**: Auto-scroll (via useSmartScroll) pauses while selection mode is active. New events still append to the stream but the viewport does not jump, preventing disruption of the user's selection. Auto-scroll resumes when selection mode exits.

### Non-Functional Requirements

- **NFR-1**: Hover action row must appear within 100ms of mouse entering the event boundary. No perceptible delay.
- **NFR-2**: Entering/exiting selection mode must not cause a visible layout shift (checkboxes should be positioned outside or overlapping the existing event layout, not pushing content sideways).
- **NFR-3**: Selection state for 5,000 events (the in-memory cap) must remain performant. No full re-render of all events on each selection toggle. Use a Set-based selection model.
- **NFR-4**: formatEventsAsMarkdown must handle events with very large content (e.g., a FileReadCard with thousands of lines) without truncation. The user explicitly selected it for copying.
- **NFR-5**: The feature must work in both stream directions (newest-at-top and newest-at-bottom). Shift+click range selection follows visual order.

### Accessibility

- **NFR-6**: Checkboxes have proper aria-label (e.g., "Select message from assistant at 2:34 PM").
- **NFR-7**: Floating action bar is keyboard-navigable (Tab cycles through buttons).
- **NFR-8**: Escape key exits selection mode from anywhere within the EventStream.
- **NFR-9**: Hover action row buttons have descriptive aria-label values and are reachable via keyboard focus (Tab into event, Tab through actions).

### Integration Points

- **EventStream** (web-components/src/components/display/EventStream.tsx): The primary integration point. Currently renders EventRenderer inside motion.div wrappers. Selection mode adds a wrapper layer around each event with checkbox + click handler + highlight styling. The existing useSmartScroll hook needs a paused parameter or equivalent to support FR-17.
- **EventRenderer** (web-components/src/components/display/EventRenderer.tsx): Currently places CopyButton statically inside TextEvent and CodeBlockWrapper. The message-level CopyButton on TextEvent is replaced by the hover action row. The code-block-level CopyButton inside CodeBlockWrapper remains (it copies just the code, not the full message).
- **Tool cards** (ShellCard, FileReadCard, GenericToolCard, etc.): Currently have inline CopyButton in their headers. These are replaced by the hover action row for message-level copy. The tool cards themselves do not need modification. The hover row wraps them from the outside.
- **CopyButton** (web-components/src/components/display/CopyButton.tsx): Continues to exist for code-block-level copy. No longer used at the message level.
- **useSmartScroll** (web-components/src/hooks/useSmartScroll.ts): Needs to accept a pause signal so auto-scroll is suppressed during selection mode.
- **ToastContext** (web-components/src/context/ToastContext.tsx): Used for "Copied N messages" feedback toast.
- **SessionPage** and **ChatPage**: Page-level components that render EventStream. They control whether selection is enabled and pass the onCopySelected callback.
- **sessionEvents.ts** (web-components/src/utils/sessionEvents.ts): The DisplayEvent type (which extends SessionEvent with toolUseCtx and settled) is the input to formatEventsAsMarkdown.

### Acceptance Criteria: Unit Tests

- **UT-1**: formatEventsAsMarkdown correctly formats a text/output event with timestamp and "Assistant" label.
- **UT-2**: formatEventsAsMarkdown correctly formats a user_input event with "User" label.
- **UT-3**: formatEventsAsMarkdown correctly formats a tool_result event with paired tool_use context (tool name, args summary, result content).
- **UT-4**: formatEventsAsMarkdown correctly formats an error event with "Error" label.
- **UT-5**: formatEventsAsMarkdown preserves code fences within markdown content.
- **UT-6**: formatEventsAsMarkdown handles multiple events in chronological order with blank-line separation.
- **UT-7**: formatEventsAsMarkdown handles a single event (used by single-message copy).

### Acceptance Criteria: Storybook Tests

- **ST-1**: Story for EventStream in default mode. Hover over an event, verify hover action row appears with Copy and Select buttons.
- **ST-2**: Story for EventStream in selection mode with mock events. Enter selection mode, select 2 events, verify selection count in action bar shows "2 selected".
- **ST-3**: Story for selection mode. Click Select All, verify all content-bearing events are checked; click Deselect All, verify all unchecked.
- **ST-4**: Story for selection mode. Click Copy, verify toast appears with "Copied N messages" message.
- **ST-5**: Story for selection mode. Press Escape, verify selection mode exits and checkboxes disappear.
- **ST-6**: Story verifying that hover action row does NOT appear on status, signal, usage, and system events.

### Acceptance Criteria: Manual Testing

- **MT-1**: Open a session with a mix of assistant text, user input, tool calls, and status events. Hover over a text event and verify action row appears. Click Copy and verify clipboard contains well-formatted markdown of that single message.
- **MT-2**: Click Select on a text event and verify selection mode activates, checkbox appears, event is highlighted. Click two more events and verify they are selected and count shows "3 selected". Click Copy and verify clipboard contains all three events as markdown.
- **MT-3**: In selection mode, click the first event, then Shift+click the fifth event. Verify events 1-5 are all selected.
- **MT-4**: In selection mode, verify the stream does not auto-scroll when new events arrive. Exit selection mode and verify auto-scroll resumes.
- **MT-5**: Toggle stream direction to newest-at-top, repeat MT-2. Verify selection and copy work correctly in reversed mode.
- **MT-6**: Right-click an event and verify context menu with Copy and Select options appears.

### Out of Scope

- **Forward button and session picker** — #979 will add the Forward action to the floating action bar and the session picker dialog.
- **"Add a note" when forwarding** — #979 scope.
- **Forwarded message envelope format and rendering** — #979 scope. Whether the envelope is JSON, markdown, or a hybrid is deferred to #979.
- **Deep link / "copy as reference link"** — future ticket. The Slack/Discord pattern of copying a link that auto-unfurls when pasted is valuable but separate from clipboard copy.
- **"Create task from output"** — future standalone ticket. The Teams-inspired pattern of creating a task from selected agent output.
- **Mobile-optimized long-press gesture** — desktop-first implementation. Mobile affordances (long-press to show context menu) can be refined in a follow-up.

### Open Questions

_None — all decisions resolved during spec review._

### PR DESCRIPTION
**Pull request by nick-pape**
## Summary
- Add Slack-style **hover action row** (Copy + Select) on each event, replacing the 13 static CopyButtons scattered across TextEvent and all 12 tool card headers
- Add Telegram-style **multi-select mode** with checkboxes, shift+click range selection, and a floating action bar (count, Select all/Deselect all, Copy, Cancel)
- New `formatEventsAsMarkdown()` utility produces labeled markdown for clipboard, reusable by the future forward feature (#979)
- Auto-scroll pauses during selection mode to prevent viewport disruption

## New components
| Component | Description |
|-----------|-------------|
| `EventHoverRow` | Wrapper with hover actions (normal) or checkbox (selection mode) |
| `FloatingActionBar` | Floating bar with bulk actions during multi-select |
| `useEventSelection` | Hook managing selection state, shift-range, clipboard copy |
| `eventContent.ts` | `isContentBearingEvent`, `getEventCopyText`, `formatEventsAsMarkdown` |

## Test plan
- [x] 28 unit tests for `eventContent.ts` (all event types, formatting, edge cases)
- [x] 12 unit tests for `useEventSelection` (enter/cancel, toggle, shift-range, select all, clipboard)
- [x] 5 Storybook stories for `EventHoverRow` with play functions
- [x] 3 Storybook stories for `FloatingActionBar` with play functions
- [x] 7 Storybook stories for `EventStream` integration (hover, selection mode, multi-select, select all, cancel)
- [x] Full `rush build` passes with no warnings
- [ ] Manual testing: hover actions, single copy, multi-select, shift-click range, floating bar, escape to cancel

Closes #978

### CODE REVIEW

File: packages/web-components/src/hooks/useEventSelection.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+  selectedIndices: ReadonlySet<number>;
+  /** Number of selected events. */
+  selectedCount: number;
+  /** Enter selection mode, optionally selecting an initial event. */
+  enterSelectionMode: (initialIndex?: number) => void;
+  /** Exit selection mode and clear all selections. */
+  cancelSelection: () => void;
+  /** Toggle an event at the given index. When shiftKey is true, selects the range from the last-toggled anchor. */
+  toggleEvent: (index: number, shiftKey?: boolean) => void;
+  /** Select all content-bearing events. */
```
[the remark this answers is not in the document]
Comment (nick-pape) [reply packages/web-components/src/hooks/useEventSelection.ts]: The events array is derived from React state via `useMemo` and is stable within a render cycle. When events change (new events arrive, grouping changes), the selection indices remain valid because selection mode pauses auto-scroll and the array only grows at the end. If the user exits and re-enters selection mode, the state is fresh. A timestamp-based ID approach would add complexity without a concrete bug to solve.

File: packages/web-components/src/utils/eventContent.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+      if (event.content.trimStart().startsWith("{")) {
+        try {
+          const parsed = JSON.parse(event.content) as Record<string, unknown>;
+          if (typeof parsed.content === "string") {
+            resultContent = parsed.content;
+          }
+        } catch { /* use as-is */ }
+      }
+      return resultContent;
+    }
```
[the remark this answers is not in the document]
Comment (nick-pape) [reply packages/web-components/src/utils/eventContent.ts]: Fixed. `getEventCopyText` now prefers `event.toolUseCtx?.detailedResult` when available, falling back to the JSON-extracted content. Added a unit test verifying this.

File: packages/web-components/src/components/display/EventHoverRow.tsx [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    (e: React.MouseEvent) => {
+      if (!isContentBearing || isSelecting) {
+        return;
+      }
+      // Only prevent default if we would show our actions
+      // For now, clicking Select from hover or context menu enters selection mode
+      // A full custom context menu is a future enhancement
+    },
+    [isContentBearing, isSelecting],
+  );
```
[the remark this answers is not in the document]
Comment (nick-pape) [reply packages/web-components/src/components/display/EventHoverRow.tsx]: Fixed. Removed the `handleContextMenu` callback entirely since it was a no-op placeholder. A full custom context menu is a future enhancement.

File: packages/web-components/src/hooks/useEventSelection.ts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  /** Copy selected events to clipboard. Returns true on success. */
+  copySelected: () => Promise<boolean>;
+}
+
+/**
+ * Hook that manages event selection state for the EventStream multi-select feature.
+ *
+ * Pure logic hook with no DOM manipulation. The caller is responsible for
+ * rendering checkboxes, highlights, and action bars based on the returned state.
+ */
```
[the remark this answers is not in the document]
Comment (nick-pape) [reply packages/web-components/src/hooks/useEventSelection.ts]: Fixed. Updated the hook TSDoc to accurately document that `copySelected` uses `navigator.clipboard` and returns `false` when unavailable.
